### PR TITLE
Refactor/moving cli to the right place

### PR DESCRIPTION
**Pull request by ryanjorgeac**

### CODE REVIEW

File: ApplicationService/connection_pool.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+class connection:
+    def cursor(self):
+        raise NotImplementedError
+
+
+class postgresql_connection_pool(connection_pool):
+    def __init__(self, createConnection=None, max_connections=1, condition=None):
+        self.create_connection = createConnection or psycopg2_create_connection
+        self.connections: dict[int, tuple] = dict()  # tuple[0] are connections and tuple[1] are conn's cursor
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:25]: tuple accept arguments to indicate type. So you must use

dict[int,tuple[connectionProtocol,cursorProtocol]]

where the *protocol types are the protocols of the generic return

File: ApplicationService/connection_pool.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        self.max_connections = max_connections
+        self.free_connections = []
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:27]: The only way to append free_connections is to get one and then refund. So, why not type then to list[tuple(connection,cursor)] and in the refund, don't ignore the cursor already created?

File: ApplicationService/connection_pool.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        id_conn = identifier.value()
+        with self.condition:
+            if id_conn not in self.connections.keys():
+                while True:
+                    if self.free_connections:
+                        conn = self.free_connections.pop()
+                        self.connections[id_conn] = (conn, conn.cursor())
+                        break
+
+                    elif len(self.connections) < self.max_connections:
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:40]: change name to self.alocatedConnections or something like that, is more semantic, dont you think?

File: ApplicationService/connection_pool.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                        conn = self.create_connection()
+                        self.connections[id_conn] = (conn, conn.cursor())
+                        break
+
+                    else:
+                        self.condition.wait()
+
+            return self.connections[id_conn][0]
+
+    def get_cursor(self, identifier: identity):
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:50]: type this method. Why is not used with a condition too?

File: ApplicationService/loginUseCase.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+logger = logging.getLogger("ApplicationService.loginUseCase")
+
+class loginUseCase:
+    def __init__(self, account_repository: internalAccountsRepository, transactional_context: transactioncontext):
+        self.account_repository = account_repository
+        self.transactional_context = transactional_context
+
+    def execute(self, username: str, password: pw):
+        logger.debug("Execute method called.")
```
Comment (petrolifero) [inline ApplicationService/loginUseCase.py:15]: remove this log

File: ApplicationService/connection_pool.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import os
+import psycopg2
+import threading
+from typing import Callable
+from ApplicationService.repositories.connectionInterface import *
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:5]: Dont use import *

File: ApplicationService/connection_pool.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+class postgresql_connection_pool(connection_pool):
+    def __init__(self, createConnection=None, max_connections=1, condition=None):
+        self.create_connection: Callable[[], connection] = createConnection or psycopg2_create_connection
+        self.allocated_connections: dict[int, tuple[connection, connection.cursor]] = dict()
+        self.max_connections: int = max_connections
+        self.free_connections: list[tuple[connection, connection.cursor]] = []
```
Comment (petrolifero) [inline ApplicationService/connection_pool.py:12]: Why the type is connection.cursor instead of cursor that you import?

File: ApplicationService/dbtransactioncontext.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+from typing import Any
+from ApplicationService.threadIdentity import identity
+from ApplicationService.connection_pool import connection_pool as cpool
+from ApplicationService.transactioncontext import transactioncontext
+
+class dbTransactionContext(transactioncontext):
+    def __init__(self, connection_pool: cpool, identifier: identity):
+        self.connection_pool = connection_pool
+        self.identifier = identifier
+        self.errors: list[Any] = []
```
Comment (petrolifero) [inline ApplicationService/dbtransactioncontext.py:10]: Can we refine this Any type? What kind of errors can a context receive?

File: ApplicationService/dbtransactioncontext.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        if not exc_val:
+            connection.commit()
+
+        else:
+            connection.rollback()
+            self.errors.append(exc_val)
+
+        self.connection_pool.refund(self.identifier)
+
+    def get_errors(self):
```
Comment (petrolifero) [inline ApplicationService/dbtransactioncontext.py:27]: type

File: ApplicationService/externalrepository.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+from ApplicationService.connection_pool import connection_pool
+from ApplicationService.threadIdentity import identity
+from ApplicationService.repositories.externalaccountsrepository import externalAccountsRepository
+
+class externalRepository(externalAccountsRepository):
+    def __init__(self, cpool: connection_pool, identifier: identity):
+        self.connection_pool = cpool
+        self.identifier = identifier
+
+    def get_by_login(self, login: str) -> maybe.maybe:
```
Comment (petrolifero) [inline ApplicationService/externalrepository.py:12]: https://docs.python.org/3/library/typing.html#typing.Generic

Can we use generics to refine the return of get_by_login?

File: ApplicationService/connection_pool.py (hunk 7 above)
Comment (ryanjorgeac) [reply ApplicationService/connection_pool.py:12]: Thank you for remembering.

File: ApplicationService/dbtransactioncontext.py (hunk 9 above)
Comment (ryanjorgeac) [reply ApplicationService/dbtransactioncontext.py:27]: The __exit__ or only get_errors' return?

Comment (petrolifero) [reply ApplicationService/dbtransactioncontext.py:27]: get_errors

File: migrations/20221101_insert_into_accounts_ryan_accounts.sql [hunk 11]
```diff
@@ -1 +1 @@
-INSERT INTO accounts (login, password) VALUES ('ryan', 'abc123');
\ No newline at end of file
+INSERT INTO accounts (login, password) VALUES ('ryan', 'c70b5dd9ebfb6f51d09d4132b7170c9d20750a7852f00680f65658f0310e810056e6763c34c9a00b0e940076f54495c169fc2302cceb312039271c43469507dc');
```
Comment (petrolifero) [inline migrations/20221101_insert_into_accounts_ryan_accounts.sql:1]: Which clean password ryan has? That kind of database operations are not schema change migrations, but fake data to test, so must be removed from migrations folder and put on test_data or something like that. After that, must be refactored to use temporary table (i.e. WITH statement) to become clean the password.

File: inputIO/inputIO.py [hunk 12]
```diff
@@ -1,11 +1,11 @@
 import getpass
 
 class inputIO:
-    def input(self,prompt):
+    def input(self, prompt: str):
```
Comment (petrolifero) [inline inputIO/inputIO.py:4]: type this methods

File: drivers/Cli/command_line_interface.py [hunk 13]
```diff
@@ -25,24 +22,19 @@ def repl(userIO, acc, transfer_use_case):
                 userIO.print(f"{e.value} is a non-positive value to transfer.")
 
 
-def main(accounts_repository, transfer_use_case, userIO):
+def main(userIO, login_use_case, transfer_use_case, open_account_use_case):
```
Comment (petrolifero) [inline drivers/Cli/command_line_interface.py:25]: type the arguments of main.

File: filemock.py [hunk 14]
```diff
@@ -9,10 +9,10 @@ def archive_with_pedro_and_his_password():
     return io.StringIO("pedro:eaf2c12742cb8c161bcbd84b032b9bb98999a23282542672ca01cc6edd268f7dce9987ad6b2bc79305634f89d90b90102bcd59a57e7135b8e3ceb93c0597117b:40\n")
```
Comment (petrolifero) [inline filemock.py:9]: that file is necessary?

File: drivers/Cli/command_line_interface.py (hunk 13 above)
Comment (ryanjorgeac) [reply drivers/Cli/command_line_interface.py:25]: That make me think about implementing a new Interface Class for UseCases, such as businessUseCase or TransactionalUseCase and do all the UseCases implements it. What do you think?

File: filemock.py (hunk 14 above)
Comment (ryanjorgeac) [reply filemock.py:9]: Not a bit.

File: ApplicationService/externalrepository.py (hunk 10 above)
Comment (ryanjorgeac) [reply ApplicationService/externalrepository.py:12]: The Generic inheritance would be in maybe class or should be used in get_by_login only?

Comment (petrolifero) [reply ApplicationService/externalrepository.py:12]: maybe

File: drivers/Cli/command_line_interface.py (hunk 13 above)
Comment (petrolifero) [reply drivers/Cli/command_line_interface.py:25]: Each Use case receive a different set of parameters, how this interface could be? I don't think that is possible

Comment (ryanjorgeac) [reply drivers/Cli/command_line_interface.py:25]: Yeah. Thinking better, there is no way to reproduce lots of UseCases using a interface. The main typing could get wrong.

Comment (ryanjorgeac) [reply drivers/Cli/command_line_interface.py:25]: Should I type the repl arguments as well?

File: migrations/20221101_insert_into_accounts_ryan_accounts.sql (hunk 11 above)
Comment (ryanjorgeac) [reply migrations/20221101_insert_into_accounts_ryan_accounts.sql:1]: I didn't understand what is a clean password. Could you please explain it better or give me some examples or documentation?

Comment (petrolifero) [reply migrations/20221101_insert_into_accounts_ryan_accounts.sql:1]: clean password = abc123. We will lost them when make that change. 

the use of WITH is postgresql-specific, so add a comment to indicate that c70b5dd9ebfb6f51d09d4132b7170c9d20750a7852f00680f65658f0310e810056e6763c34c9a00b0e940076f54495c169fc2302cceb312039271c43469507dc is the hash of abc123.

Then make yoyo down to remove your migrations from database. Then move that migration to a folder "fake_data". Now reapply migrations. Now you can apply the fake data outside migrations, so when we put on production, the fake data will not be there.

File: maybe.py [hunk 15]
```diff
@@ -1,39 +1,47 @@
...
+
+class just(maybe[T]):
+    def __init__(self, value: T):
         self.value = value
 
     def map(self, function):
         return just(function(self.value))
 
-    def orElse(self, default):
+    def orElse(self, default: T):
```
Comment (petrolifero) [inline maybe.py:24]: tyoe the return of the methods.

File: maybe.py [hunk 16]
```diff
@@ -1,39 +1,47 @@
...
         return default()
 
     def orElseThrow(self, value):
         raise value
 
+
 def getFirstNotEmpty(possible_not_empties):
```
Comment (petrolifero) [inline maybe.py:43]: type the function

File: ApplicationService/repositories/internalaccountsrepository.py [hunk 17]
```diff
@@ -0,0 +1,11 @@
+from password import password as pw
+
+class internalAccountsRepository:
+    def add_account(self, login: str, password: pw):
```
Comment (petrolifero) [inline ApplicationService/repositories/internalaccountsrepository.py:4]: type return of methods

File: ApplicationService/repositories/externalaccountsrepository.py [hunk 18]
```diff
@@ -0,0 +1,6 @@
+class externalAccountsRepository:
+    def get_by_login(self, login: str):
```
Comment (petrolifero) [inline ApplicationService/repositories/externalaccountsrepository.py:2]: type return of methods

File: test_internalAccount.py [hunk 19]
```diff
@@ -1,31 +1,32 @@
-import externalAccount
-import internalAccount
+from ApplicationService.external_account import externalAccount, negativeIncrementException
+from ApplicationService.internal_account import internalAccount, insufficientFundsException
+import password
```
Comment (petrolifero) [inline test_internalAccount.py:3]: Why password is no imported as the others? Like from password import password?

File: inputIO/inputIO.py [hunk 20]
```diff
@@ -1,11 +1,11 @@
-import getpass
+from getpass import getpass
 
 class inputIO:
-    def input(self,prompt):
+    def input(self, prompt: str) -> input:
```
Comment (petrolifero) [inline inputIO/inputIO.py:4]: input type?

File: drivers/Web/test_bankApplication.py [hunk 21]
```diff
@@ -32,12 +32,12 @@ def test_request_root_resource_is_html():
     bankApplicationReturn = bankApplication.root(request)
     assert bankApplicationReturn.getHeaders()["Content-Type"] == "text/html"
 
-def test_post_status_200():
+def test_post_status_303():
     header = {}
     resource = httpRequest.http_resource("/", {})
     method = "POST"
-    body = b"login=ryanbanco&password=abc123"
+    body = b"_login=ryanbanco&password=abc123"
```
Comment (petrolifero) [inline drivers/Web/test_bankApplication.py:39]: Why this change??

File: drivers/Web/routes.py [hunk 22]
```diff
@@ -1,39 +1,41 @@
 import maybe
```
Comment (petrolifero) [inline drivers/Web/routes.py:1]: why not import only maybe class as others?

File: drivers/Web/router.py [hunk 23]
```diff
@@ -1,15 +1,15 @@
...
+from drivers.Web.httpRequest import httpRequest
 
 logger = logging.getLogger("drivers.Web.router")
 
 class router:
     def __init__(self, *routes):
         self.routes = routes
 
-    def __call__(self, request):
+    def __call__(self, request: httpRequest):
```
Comment (petrolifero) [inline drivers/Web/router.py:12]: why not return?

File: drivers/Web/router.py [hunk 24]
```diff
@@ -1,15 +1,15 @@
 import logging
-
 import maybe
 from drivers.Web import httpResponse
+from drivers.Web.httpRequest import httpRequest
 
 logger = logging.getLogger("drivers.Web.router")
 
 class router:
     def __init__(self, *routes):
```
Comment (petrolifero) [inline drivers/Web/router.py:9]: type the input. You can use

def __init__(self,*routes : route)

File: drivers/Web/bankApplication.py [hunk 25]
```diff
@@ -24,17 +24,17 @@ def get(self, request: httpRequest.httpRequest) -> httpResponse.httpResponse:
 
         return response
 
-    def post(self, request):
+    def post(self, request: httpRequest.httpRequest) -> httpResponse.httpResponse:
```
Comment (petrolifero) [inline drivers/Web/bankApplication.py:27]: import only the class needed, to follow your standard

File: ApplicationService/transactioncontext.py [hunk 26]
```diff
@@ -0,0 +1,9 @@
+class transactioncontext:
+    def __enter__(self):
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        raise NotImplementedError
+
+    def get_errors(self):
```
Comment (petrolifero) [inline ApplicationService/transactioncontext.py:8]: can you return that List of BusinnessError, right?

File: ApplicationService/threadIdentity.py [hunk 27]
```diff
@@ -0,0 +1,8 @@
+import threading
+from ApplicationService.repositories.identity import identity
+
+class thread_identity(identity):
+    def value(self) -> int:
+        return threading.current_thread().ident
```
Comment (petrolifero) [inline ApplicationService/threadIdentity.py:6]: import only the function you need

File: ApplicationService/test_connection_pool.py [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+
+def test_connection_pool_refund_trying_to_create_conn_but_max_has_reached():
+    psql_conn = connection_pool.postgresql_connection_pool(fake_connection)
+    id1 = fake_identity(141)
+    id2 = fake_identity(555)
+
+    def get_conn(id_conn):
+        return psql_conn.get_connection(id_conn)
+
+    thread1 = threading.Thread(target=get_conn, args=(id1,))
```
Comment (petrolifero) [inline ApplicationService/test_connection_pool.py:96]: import only class needed

File: ApplicationService/repositories/connectionInterface.py [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+        raise NotImplementedError
+
+    def rollback(self):
+        raise NotImplementedError
+
+class connection_pool:
+    def get_connection(self, identifier: identity) -> connection:
+        raise NotImplementedError()
+
+    def get_cursor(self, identifier: identity):
```
Comment (petrolifero) [inline ApplicationService/repositories/connectionInterface.py:25]: type return

File: inputIO/inputIO.py (hunk 20 above)
Comment (ryanjorgeac) [reply inputIO/inputIO.py:4]: so i need to return the input's return?

File: drivers/Web/test_bankApplication.py (hunk 21 above)
Comment (ryanjorgeac) [reply drivers/Web/test_bankApplication.py:39]: Unfortunatelly, this happened because of a refactor in a login campus that should only change attribute name and its usages but renamed even comments and strings. Already solved.

File: drivers/Web/routes.py (hunk 22 above)
Comment (ryanjorgeac) [reply drivers/Web/routes.py:1]: Because I also use maybe.just and maybe.nothing. Should I import those 3 methods using "from maybe import maybe, just, nothing"?

File: ApplicationService/transactioncontext.py (hunk 26 above)
Comment (ryanjorgeac) [reply ApplicationService/transactioncontext.py:8]: correctly

File: inputIO/inputIO.py (hunk 20 above)
Comment (ryanjorgeac) [reply inputIO/inputIO.py]: in that case, -> str ?

File: drivers/Web/routes.py (hunk 22 above)
Comment (petrolifero) [reply drivers/Web/routes.py:1]: yes

File: inputIO/inputIO.py [hunk 30]
```diff
@@ -1,11 +1,11 @@
...
+from getpass import getpass
 
 class inputIO:
-    def input(self,prompt):
+    def input(self, prompt: str) -> str:
         return input(prompt)
 
-    def inputoccult(self,prompt):
-        return getpass.getpass(prompt)
+    def inputoccult(self, prompt: str) -> getpass:
```
Comment (petrolifero) [inline inputIO/inputIO.py:7]: return getpass?

File: ApplicationService/internalrepository.py [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+        cursor = self.connection_pool.get_cursor(self.identifier)
+        cursor.execute("SELECT * FROM accounts WHERE login = %s ", (new_login,))
+        account_query_result = cursor.fetchone()
+
+        if not account_query_result:
+            cursor.execute("INSERT INTO accounts (login,password,balance) VALUES (%s,%s,%s)", (new_login, str(new_password), 0))
+
+        return not account_query_result
+
+    def authentication(self, login: str, password: pw) -> maybe.maybe:
```
Comment (petrolifero) [inline ApplicationService/internalrepository.py:26]: import maybe class and make them specific maybe[internalAccount]

File: ApplicationService/externalrepository.py [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+from ApplicationService.connection_pool import connection_pool
+from ApplicationService.threadIdentity import identity
+from ApplicationService.repositories.externalaccountsrepository import externalAccountsRepository
+
+class externalRepository(externalAccountsRepository):
+    def __init__(self, cpool: connection_pool, identifier: identity):
+        self.connection_pool = cpool
+        self.identifier = identifier
+
+    def get_by_login(self, login: str) -> maybe.maybe[externalAccount]:
```
Comment (petrolifero) [inline ApplicationService/externalrepository.py:12]: import only maybe class

**Review by petrolifero — APPROVED** [review]
